# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
#normal function using def

def add_two(num):
    return num + 2

add_two(4)

6

In [2]:
#lambda function

(lambda num: num + 2)(4)

6

In [3]:
(lambda num: num **2)(4)

16

In [ ]:
result = (lambda num: num -3)(2)
print(result)

#### Saving to a Variable

In [7]:
square_nums = lambda x: x**2

In [8]:
square_nums(2)

4

In [9]:
square_nums(3)

9

In [10]:
square_nums(4)

16

#### Multiple Inputs

In [11]:
(lambda num1, num2, num3: num1 + num2 + num3)(4,5,6)

15

#### Passing a Lambda into a Function

In [12]:
def double_numbers(a_list):
    for i in range(len(a_list)):
        a_list[i] = a_list[i] * 2
        
    return a_list

double_numbers([1,2,3,4,5])

[2, 4, 6, 8, 10]

In [13]:
def half_numbers(a_list):
    for i in range(len(a_list)):
        a_list[i] = a_list[i] / 2
        
    return a_list

half_numbers([1,2,3,4,5])

[0.5, 1.0, 1.5, 2.0, 2.5]

In [14]:
def increment_numbers(a_list):
    for i in range(len(a_list)):
        a_list[i] = a_list[i] + 1
        
    return a_list

increment_numbers([1,2,3,4,5])

[2, 3, 4, 5, 6]

In [15]:
def modify_numbers(a_list, f):
    for i in range(len(a_list)):
        a_list[i] = f(a_list[i])
        
    return a_list

In [17]:
#Double numbers
modify_numbers([1,2,3,4,5], lambda x: x * 2)

#half numbers
modify_numbers([1,2,3,4,5], lambda x: x / 2)

#increment numbers
modify_numbers([1,2,3,4,5], lambda x: x + 1)

#square numbers, subtract 1, add 4
modify_numbers([1,2,3,4,5], lambda x: (x **2) - 1 + 4)

[4, 7, 12, 19, 28]

#### Returning a Lambda from a Function

In [19]:
lambda num1, num2: (num1 + num2 + 4)**2

<function __main__.<lambda>(num1, num2)>

In [20]:
lambda num1, num2: (num1 + num2 + 7)**3

<function __main__.<lambda>(num1, num2)>

In [21]:
def build_lambda(adder, power):
    return lambda num1, num2: (num1 + num2 + adder) **power

In [26]:
(build_lambda(7,3))(4,5)



4096

In [25]:
(build_lambda(4,2))(4,5)

169

#### If Statements within Lambdas

In [28]:
# square every number in a list
a_list = [1,2,3,4,5]

modify_numbers(a_list, lambda x: x ** 2 if x % 2 == 1 else x)

[1, 2, 9, 4, 25]

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [33]:
f = lambda x: x**3

print(f(6))

216


In [32]:
print(f(5))

125


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [35]:
# map(function to be used, list to be used)

a_list = [1,2,3,4,5]

print(list(map(lambda x: x *2, a_list)))
print(list(map(lambda x: x *3, a_list)))
print(list(map(lambda x: x *4, a_list)))
print(list(map(lambda x: x **2, a_list)))

[2, 4, 6, 8, 10]
[3, 6, 9, 12, 15]
[4, 8, 12, 16, 20]
[1, 4, 9, 16, 25]


#### Using Lambda's with Map

In [ ]:
print(list(map(lambda x: x *2, a_list)))

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [37]:
print(list(map(lambda x: (x *2) -1, a_list)))

[1, 3, 5, 7, 9]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [39]:
a_list = [1,2,3,4,5,6,7,8,9,10]

list(filter(lambda num: num % 2 == 0, a_list))

[2, 4, 6, 8, 10]

In [97]:
my_list = list('a,b,c')
my_list

['a', ',', 'b', ',', 'c']

In [99]:
my_filtered_list = list(filter(lambda x: x != ',', my_list))

In [100]:
my_filtered_list

['a', 'b', 'c']

#### Using Lambda's with Filter()

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [45]:
from statistics import mean

a_list = [3,4,5,6,7,8,9,10]

print(mean(a_list))

list(filter(lambda num: num >= mean(a_list), a_list))

6.5


[7, 8, 9, 10]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [47]:
from functools import reduce

# use reduce to find a sum of a list

a_list = [1,2,3,4,5]

def add_nums(num1, num2):
    print(num1, num2)
    return num1 + num2

reduce(add_nums, a_list)


1 2
3 3
6 4
10 5


15

In [49]:
def multiply_nums(num1, num2):
    return num1 * num2

reduce(multiply_nums, a_list)

120

In [50]:
reduce(lambda num1, num2: num1 * num2, a_list)

120

#### Using Lambda's with Reduce()

In [ ]:
reduce(lambda num1, num2: num1 * num2, a_list)

#### In-Class Exercise #4 <br>
<p>Use the reduce function to subtract the numbers in the list below together with a lambda function.</p>

In [51]:
a_list = [1,2,3,4,5]

# 1-2-3-4-5

reduce(lambda num1, num2: num1 - num2, a_list)

-13

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [55]:
def factorial(number):
    # base case
    if number <= 1:
        return number
    return number * factorial(number -1)

In [56]:
factorial(5)

120

In [59]:
def subtraction(number):
    # base case
    if number <= 1:
        return number
    return number - subtraction(number -1)

#### Writing a Factorial Function

3

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [62]:
def subtraction(number):
    # base case
    if number <= 1:
        print('recursive_subtraction(1) = 1')
        return number
    
    print(f'recursive_subtraction({number}) = {number} - recursive_subtraction({number - 1})')
    return number - subtraction(number -1)

In [63]:
subtraction(5)

recursive_subtraction(5) = 5 - recursive_subtraction(4)
recursive_subtraction(4) = 4 - recursive_subtraction(3)
recursive_subtraction(3) = 3 - recursive_subtraction(2)
recursive_subtraction(2) = 2 - recursive_subtraction(1)
recursive_subtraction(1) = 1


3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [64]:
my_nums = list(range(1,101))

In [68]:
def sqrt_nums(a_list):
    for num in a_list:
        yield num**.5
    



In [89]:
my_nums_sqrt = sqrt_nums(my_nums)

In [90]:
next(my_nums_sqrt)

1.0

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [139]:
# for i in numbers_squared(10)

def num_sq(stop,start,step = 1):
    while start >= stop:
        yield start ** 2
        start -= step
        
for i in num_sq(0,start = 3):
    print(i)

9
4
1
0


In [95]:
def squared_nums(a_list):
    result = []
    for num in a_list:
        result.append(num**2)
        
    return result

squared_nums(my_nums)

[1,
 4,
 9,
 16,
 25,
 36,
 49,
 64,
 81,
 100,
 121,
 144,
 169,
 196,
 225,
 256,
 289,
 324,
 361,
 400,
 441,
 484,
 529,
 576,
 625,
 676,
 729,
 784,
 841,
 900,
 961,
 1024,
 1089,
 1156,
 1225,
 1296,
 1369,
 1444,
 1521,
 1600,
 1681,
 1764,
 1849,
 1936,
 2025,
 2116,
 2209,
 2304,
 2401,
 2500,
 2601,
 2704,
 2809,
 2916,
 3025,
 3136,
 3249,
 3364,
 3481,
 3600,
 3721,
 3844,
 3969,
 4096,
 4225,
 4356,
 4489,
 4624,
 4761,
 4900,
 5041,
 5184,
 5329,
 5476,
 5625,
 5776,
 5929,
 6084,
 6241,
 6400,
 6561,
 6724,
 6889,
 7056,
 7225,
 7396,
 7569,
 7744,
 7921,
 8100,
 8281,
 8464,
 8649,
 8836,
 9025,
 9216,
 9409,
 9604,
 9801,
 10000]

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [138]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

def no_empty_strings(a_list):

    my_filtered_list = list(filter(lambda x: x >=',', places))
    return my_filtered_list
    
no_empty_strings(places)


['Argentina', 'San Diego', 'Boston', 'New York']

In [124]:
#my_filtered_list

['Argentina', 'San Diego', 'Boston', 'New York']

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [136]:
authors = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

def authors_last(a_list):
    
    authors.sort(key = lambda name: name.split(" ")[-1].lower())
    return authors

authors_last(authors)

['Victor aNisimov',
 'Gary A.J. Bernstein',
 'Joel Carter',
 'Andrew P. Garfield',
 'David hassELHOFF']

### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [137]:
# F = (9/5)*C + 32
# print(list(map(lambda x: x *2, a_list)))

places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

def Farhenheit(a_list):
    
    places_f = list(map(lambda c: (c[0], (9/5)*c[1] +32), places))
    return places_f

Farhenheit(places)


[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [135]:
# def subtraction(number):
#     # base case
#     if number <= 1:
#         return number
#     return number - subtraction(number -1)

def Sequence(num):
    if num == 0:
        return 0
    elif num == 1:
        return 1
    else:
        return Sequence(num - 1) + Sequence(num - 2)
    
def fib(number):
    for i in range(number + 1):
        print("Iteration({i}): {x}".format(i=i,x=Sequence(i+1)))

fib(5)
    

Iteration(0): 1
Iteration(1): 1
Iteration(2): 2
Iteration(3): 3
Iteration(4): 5
Iteration(5): 8
